async_task.go

```diff
@@ -88,12 +90,12 @@ func (t *TaskStatus) Wait(ctx context.Context) (interface{}, error) {
 	case <-ch:
 		return t.result, t.err
 	case <-ctx.Done():
-		t.finish(StateCanceled, nil, ErrTimeout)
-		return t.result, t.err
+		return nil, ErrWaitTimeout
 	}
 }
 
 // WaitWithTimeout block current thread/routine until task finished or failed, or exceed the duration specified.
+// timeout only stop waiting, taks will remain running.
 func (t *TaskStatus) WaitWithTimeout(ctx context.Context, timeout time.Duration) (interface{}, error) {
 	// return immediately if task already in terminal state.
 	if t.state.IsTerminalState() {
```

Comment:
**question**: this is already implemented in Wait(...) do you need it here?

Reply:
I can't comment on the below lines, but, you are calling `defer cancelFunc()` that's also called in the Wait() implementation, is shouldn't that all be just handled in Wait()?

Reply:
otherwise the cancelfunc is called twice? maybe write a test to check that assumption?

Reply:
this defer cancelFunc() is the cancelFunc from WithTimeout.

in Wait the cancelFunc is the context started the task